datastore/transaction.js

```diff
@@ -0,0 +1,317 @@
+    } else {
+      // The transaction completed successfully.
+      callback();
+    }
+  });
+}
+// [END transactional_update]
+
+function Transaction(projectId) {
+  this.datastore = gcloud.datastore({
+    projectId: projectId
+  });
+
+  this.fromKey = this.datastore.key(['Bank', 1, 'Account', 1]);
+  this.toKey = this.datastore.key(['Bank', 1, 'Account', 2]);
+
+  this.originalBalance = 100;
+  this.amountToTransfer = 10;
+}
+
+Transaction.prototype.restoreBankAccountBalances = function(config, callback) {
+  var saveArray = config.keys.map(function(key) {
+    return {
+      key: key,
+      data: {
+        balance: config.balance
+      }
+    };
+  });
+
+  this.datastore.save(saveArray, callback);
+};
+
+Transaction.prototype.testTransactionalUpdate = function(callback) {
+  var self = this;
+
+  var fromKey = this.fromKey;
+  var toKey = this.toKey;
+  var originalBalance = this.originalBalance;
+  var amountToTransfer = this.amountToTransfer;
+
+  this.restoreBankAccountBalances({
+    keys: [fromKey, toKey],
+    balance: originalBalance
+  }, function(err) {
+    if (err) {
+      callback(err);
+      return;
+    }
+
+    // Overwrite so the real Datastore instance is used in `transferFunds`.
+    var datastoreMock = datastore;
+    datastore = self.datastore;
+
+    transferFunds(fromKey, toKey, amountToTransfer, function(err) {
+      // Restore `datastore` to the mock API.
+      datastore = datastoreMock;
+
+      if (err) {
+        callback(err);
+        return;
+      }
+
+      self.datastore.get([
+        fromKey,
+        toKey
+      ], function(err, accounts) {
+        if (err) {
+          callback(err);
+          return;
+        }
+
+        var transactionWasSuccessful =
+          accounts[0].data.balance === originalBalance - amountToTransfer &&
+          accounts[1].data.balance === originalBalance + amountToTransfer;
+
+        if (!transactionWasSuccessful) {
+          callback(new Error('Accounts were not updated successfully.'));
+        } else {
+          callback();
+        }
+      });
+    });
+  });
+};
+
+Transaction.prototype.testTransactionalRetry = function(callback) {
+  // Overwrite so the real Datastore instance is used in `transferFunds`.
+  var datastoreMock = datastore;
+  datastore = this.datastore;
+
+  var originalCallback = callback;
+  callback = function() {
+    // Restore `datastore` to the mock API.
+    datastore = datastoreMock;
+    originalCallback.apply(null, arguments);
+  };
+
+  var fromKey = this.fromKey;
+  var toKey = this.toKey;
+
+  this.restoreBankAccountBalances({
+    keys: [fromKey, toKey],
+    balance: this.originalBalance
+  }, function(err) {
+    if (err) {
+      callback(err);
+      return;
+    }
+
+    // [START transactional_retry]
+    var async = require('async');
+
+    function attemptTransfer(callback) {
+      transferFunds(fromKey, toKey, 10, callback);
+    }
+
+    async.retry(5, attemptTransfer, callback);
```

Comment:
Nope seems ok. I'd rather show the broad concepts here. If there are more odd cases I think they should move into the gcloud-node docs.